qiskit/circuit/__init__.py

```diff
@@ -832,10 +832,10 @@
 method <https://numpy.org/devdocs/user/basics.interoperability.html#the-array-method>`__,
 ``__array__``.  This is used by :meth:`Gate.to_matrix`, and has the signature:
 
-.. currentmodule:: None
+.. currentmodule:: builtins
```

Comment:
Why is this changed?  `__array__` is not a Python builtin and isn't something that's part of the `builtins` module - it doesn't have a module, which is spelled `currentmodule:: None` in Sphinx.

Sphinx might be generating a modified `id` to avoid HTML problems, but if that's causing problems, it's on the scripts that are parsing the output of Sphinx to fix that, not to change the semantics of the documentation.  The Sphinx-generated HTML entries do correctly name this method `__array__`, and it appears correctly in `objects.inv`.

If it's _really_ impossible to fix the conversion scripts to use Sphinx data (such as using `objects.inv` as a source of truth) rather than trying to scrape it from escaped HTML, then potentially the cleaner workaround is to document this as being a potential method on `object` (which _is_ in `builtins`) or somewhat lying and documenting it as part of `Instruction`.

Reply:
I have been testing the use of `objects.inv` to determine the headers instead of reading the HTML and it would be more tricky than it seems. For example, for some classes, we have duplicated IDs that reference a rust 'class' and a python class:

```
py:class
    qiskit._accelerate.qasm3.CustomGate      apidoc/qasm3.html#qiskit.qasm3.CustomGate
    qiskit.qasm3.CustomGate                  apidoc/qasm3.html#qiskit.qasm3.CustomGate
```

I'm not sure if we can always differentiate between them, so maybe it's better to go with the workaround as `__array__` is the only instance of this problem that we have found. In the last commit, I changed `builtins` to `object` as you suggested :+1:

Reply:
Fwiw, the `objects.inv` format also includes a "priority" field, which might be what distinguishes the canonical reference?  I'm not super sure.  For example (the `printf` is just appending the magic numbers for zlib compression, and the `tail` cuts out the Sphinx comment header):
```bash
{ printf "\x1f\x8b\x08\x00\x00\x00\x00\x00" ; tail -n+5 docs/_build/html/objects.inv ; } | gzip -d | grep 'CustomGate'
```
```text
qiskit._accelerate.qasm3.CustomGate py:class -1 apidoc/qasm3.html#qiskit.qasm3.CustomGate -
qiskit.qasm3.CustomGate py:class 1 apidoc/qasm3.html#$ -
```
it's the -1 / 1 that I'm talking about.  But that might be something different too.

Anyway - thanks for the change.  What I meant was ever so slightly different: leaving `.. currentmodule:: None`, but documenting the method as `object.__array__`.  I've updated the PR to be like that, and I checked that it appears like that in the HTML's `id` tag.

Reply:
The generated HTML is this, for reference:
```html
<dl class="py method">
<dt class="sig sig-object py" id="object.__array__">
<span class="sig-prename descclassname"><span class="pre">object.</span></span><span class="sig-name descname"><span class="pre">__array__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#object.__array__" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a Numpy array representing the gate. This can use the gate’s
<a class="reference internal" href="../stubs/qiskit.circuit.Instruction.html#qiskit.circuit.Instruction.params" title="qiskit.circuit.Instruction.params"><code class="xref py py-attr docutils literal notranslate"><span class="pre">params</span></code></a> field, and may assume that these are numeric
values (assuming the subclass expects that) and not
<a class="reference internal" href="#circuit-compile-time-parameters"><span class="std std-ref">compile-time parameters</span></a>.</p>
<p>For greatest efficiency, the returned array should default to a dtype of <a class="reference external" href="https://docs.python.org/3/library/functions.html#complex" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">complex</span></code></a>.</p>
</dd></dl>
```